readme.org
#+Title: Sunday Vespers, as used at St. Joseph's, Gateshead

This repository holds vespers booklets suitable for congregation and choir,
according to the sadly limited rubrics of the 1962 books (i.e. antiphons are not
doubled).

This repository is semi-automated.  For more information see the [[https://st-josephs-gateshead.github.io/SundayVespers][documentation]].


CHANGELOG.md
# [0.3.0](https://github.com/St-Josephs-Gateshead/SundayVespers/compare/v0.2.1...v0.3.0) (2022-03-05)


### Features

* dominica I in quad. ([b4e9360](https://github.com/St-Josephs-Gateshead/SundayVespers/commit/b4e9360e21bfa3dcb84910b082c36710e69c6a20))

## [0.2.1](https://github.com/St-Josephs-Gateshead/SundayVespers/compare/v0.2.0...v0.2.1) (2021-11-14)


### Bug Fixes

* **adventII:** correct psalms ([87759c7](https://github.com/St-Josephs-Gateshead/SundayVespers/commit/87759c777480ddbec265b1cffa8056d93bd90d31))

# [0.2.0](https://github.com/St-Josephs-Gateshead/SundayVespers/compare/v0.1.2...v0.2.0) (2021-11-06)


### Features

* add xix after pentecost, fixing psalms ([910f2a1](https://github.com/St-Josephs-Gateshead/SundayVespers/commit/910f2a11ac7c5057cbaba10913f3cf6751596153))

# [0.1.0](https://github.com/St-Josephs-Gateshead/SundayVespers/compare/v0.0.13...v0.1.0) (2021-09-19)


### Features

* use giplain throughout ([d6bf7cf](https://github.com/St-Josephs-Gateshead/SundayVespers/commit/d6bf7cf369a4c72c0b4fe952619b8c862d9676d5))
* **vespers:** add xvii after pentecost ([667db44](https://github.com/St-Josephs-Gateshead/SundayVespers/commit/667db449ced4c340660fa862600019c8e80c60f3))


psalm-parser.py
#!/usr/bin/python3
"""Basic parser for psalms as downloaded and converted to tex.
Format is:
{XX~} half verse * half verse
continue half verse

Last two verses are Gloria.
"""

import argparse

parser = argparse.ArgumentParser()
parser.add_argument("PATH", help="File to read in.")
parser.add_argument("-n", help="psalm number to output")
parser.add_argument("--range", help="Range to print")
parser.add_argument("-g", "--gloria", help="Print Gloria?", action="store_true")
args = parser.parse_args()

nos = [
    "Zero",
    "One",
    "Two",
    "Three",
    "Four",
    "Five",
    "Six",
    "Seven",
    "Eight",
    "Nine",
    "Ten",
]

verses = {}

with open(args.PATH, "r") as f:
    for line in f:
        if line.startswith("{"):
            no, vpart = line.split("}", 1)
            no = int(no.lstrip("{").strip(".~"))
            verses[no] = (
                vpart.strip("\n")
                .replace("~†", "~\\+\\\\\n \\vin")
                .replace("~*", "~\\*\\\\\n \\vin")
            )
            last_verse = no
        else:
            verses[last_verse] = (
                verses[last_verse]
                + " "
                + line.strip("\n")
                .replace("~†", "~\\+\\\\\n \\vin")
                .replace("~*", "~\\*\\\\\n \\vin")
            )

if not args.range:
    start = min(verses.keys())
    stop = max(verses.keys()) - 2
else:
    start, stop = args.RANGE.split("-")

if args.n:
    print(f"\\newcommand{{\\psalm{nos[int(args.n)]}}}{{")

for i in range(int(start), int(stop) + 1):
    print(verses[i])

if args.gloria:
    for i in list(verses.keys())[-2:]:
        print(verses[i])

if args.n:
    print("}\n\n")


psalm_parser.py
#!/usr/bin/python3
"""Basic parser for psalms as downloaded and converted to tex.
Format is:
{XX~} half verse * half verse

Last two verses are Gloria.
"""

from collections import namedtuple
from pathlib import Path
from re import search, sub

from jinja2 import Environment, Template

env = Environment(variable_start_string="{*", variable_end_string="*}")

ab_line_template = env.from_string(
    "\\no{{*no*}}{% if flex %}\\{*ab*} {*flex*}~+\\\\\n{%endif%}\\{*ab*} {*start*}~\*\\\\\n\\{*ab*} {*end*}\\\\\n"
)

Line = namedtuple("Line", ["flex", "start", "end"])

indented_line_template = Template(
    "{% if flex %}\\vin {{flex}}~+\\\\\n{%endif%}\\vin {{start}}~\*\\\\\n\\vin {{end}}\\\\\n"
)
unindented_line_template = Template(
    "{% if flex %}{{flex}}~+\\\\\n{%endif%}{{start}}~\*\\\\\n{{end}}\\\\\n"
)
unindent_first_line_template = Template(
    "{% if flex %}{{flex}}~+\\\\\n\\vin {%endif%}{{start}}~\*\\\\\n\\vin {{end}}\\\\\n"
)


def process_verses(inf: Path, indent_mode: str):

    verses = {}
    with inf.open() as f:
        versetoggle = False
        for line in f:
            no = int(search(r"^{([0-9]+)", line).group(1))
            line = sub(r"^{.+?} *", "", line)
            match = search(r"(.+)~†(.+)~\* *(.+)\\\\$", line)
            if match:
                line = dict(
                    flex=match.group(1), start=match.group(2), end=match.group(3)
                )
            else:
                match = search(r"(.+)~\* *(.+)\\\\$", line)
                if not match:
                    with Path("/tmp/failed").open("a") as f:
                        f.write(f"{inf}: {line}\n")
                    assert False, "Failed to match"
                line = dict(flex=None, start=match.group(1), end=match.group(2))

            if indent_mode == "alternate":
                if versetoggle:
                    verses[no] = indented_line_template.render(line)
                else:
                    verses[no] = unindented_line_template.render(line)
            elif indent_mode == "hangline":
                verses[no] = unindent_first_line_template.render(line)
            elif indent_mode == "ab":
                line["ab"] = "oddverse" if no % 2 else "evenverse"
                line["no"] = no
                verses[no] = ab_line_template.render(line)
            else:
                verses[no] = unindented_line_template.render(line)
            versetoggle = not versetoggle
    return verses


def main():
    import argparse
    import warnings

    parser = argparse.ArgumentParser()
    parser.add_argument("ANTIPHON", help="Antiphon file to parse", type=Path)
    parser.add_argument("PSALM_NAME", help="psalm name to output")
    parser.add_argument("--range", help="Range to print")
    parser.add_argument("-ng", "--no-gloria", help="Omit Gloria?", action="store_true")
    parser.add_argument("-nc", "--no-copy", action="store_true", help="Skip .gabc file")
    parser.add_argument(
        "--indent-mode",
        help="Indentation mode, one of 'alternate', 'hangline' or 'none'.",
        default="alternate",
    )
    # parser.add_argument("--line-numbers", action="store_true", help="Show line numbers.")
    args = parser.parse_args()

    nos = [
        "Zero",
        "One",
        "Two",
        "Three",
        "Four",
        "Five",
        "Six",
        "Seven",
        "Eight",
        "Nine",
        "Ten",
    ]

    antiphon = args.ANTIPHON.expanduser()
    mode, termination = None, None
    with antiphon.open() as f:
        for line in f.readlines():
            if match := search("mode:(.+);", line):
                mode = match.group(1)
            if match := search("mode-differentia:(.+);", line):
                termination = match.group(1)
    if not mode:
        raise Exception("No mode found")
    if not termination:
        warnings.warn("No termination found")

    if match := search("ant([0-9])", antiphon.stem):
        logical_name = match.group(1)
    elif match := search("(.+)-ant", antiphon.stem):
        logical_name = match.group(1)
    else:
        raise Exception("Unable to determine logical name")

    try:
        int(args.PSALM_NAME)
    except ValueError:
        args.PSALM_NAME = args.PSALM_NAME.title()

    if termination:
        termination = termination.replace("*", "star")
        inf = Path(f"../../psalms/{args.PSALM_NAME}-{mode}{termination}.tex")
    else:
        inf = Path(f"../../psalms/{args.PSALM_NAME}-{mode}.tex")

    verses = process_verses(inf, args.indent_mode)

    if args.range:
        start, stop = (int(x) for x in args.RANGE.split("-"))
    else:
        start = min(verses.keys())
        stop = max(verses.keys()) - 2

    outd = Path(".")
    try:
        logical_name = int(logical_name)
        outf = outd / f"psalm{logical_name}.gabc"
        tex_name = f"\\newcommand{{\\psalm{nos[int(logical_name)]}}}{{\n"
    except ValueError:
        outf = outd / f"{logical_name}.gabc"
        tex_name = f"\\newcommand{{\\{logical_name}}}{{\n"

    with (outd / "psalms.tex").open("a") as f:
        f.write(tex_name)

        for i in range(start, stop + 1):
            f.write(verses[i] + "\n")

        if not args.no_gloria:
            for i in list(verses.keys())[-2:]:
                f.write(verses[i] + "\n")

        f.write("}\n\n")

    if not args.no_copy:
        with outf.open("w") as f:
            with inf.with_suffix(".gabc").open() as g:
                f.write(g.read())


if __name__ == "__main__":
    main()


compline/get_psalms.py
#!/usr/bin/python3
"""Script to get all psalms for day in right tone and drop them in directory with right names"""
from pathlib import Path
from shutil import copy
from argparse import ArgumentParser

src = Path("/home/john/Library/Music/St-Cuthberts-Liturgy/gregorian-psalms/psalms")

parser = ArgumentParser()

parser.add_argument("PSALMS", nargs="+", help="psalm")
parser.add_argument("TONE", help="tone")
parser.add_argument("DAY", help="day")
parser.add_argument("--destdir", help="destdir, default = .")

args = parser.parse_args()
if not args.destdir:
    args.destdir = Path(".")
else:
    args.destdir = Path(args.destdir)

for i, psalm in enumerate(args.PSALMS):
    i+=1
    copy(src / f"{psalm}-{args.TONE}.gabc", args.destdir / f"{args.DAY}{i}.gabc")
    copy(src / f"{psalm}-{args.TONE}.tex", args.destdir / f"{args.DAY}{i}.tex")
    




compline/generate_palterium.py
#!/usr/bin/python3
from pathlib import Path

days = {
    "Dominica": [4, 90, 133],
    "Feria Secunda": [6, 7.1, 7.2],
    "Feria Tertia": [11, 12, 15],
    "Feria Quarta": [33.1, 33.2, 60],
    "Feria Quinta": [69, 70.1, 70.2],
    "Feria Sexta": [76.1, 76.2, 85],
    "Sabbato": [87, 102.1, 102.2]
}

output = Path("psalterium.tex")
if output.exists():
    output.rename("psalterium.tex.bak")

psalterium = """

\\section{Psalterium}
\\label{sec:psalterium}

\\setlength{\\vindent}{\\vgap}

\\let\\tmp\\subsection
\\renewcommand{\\subsection}{\\clearpage\\tmp}

"""

psalm_parts = {"1": "i", "2": "ii"}

for day, psalms in days.items():
    short_day = day.split(" ")[-1].lower()[:3]
    psalterium += f"\\subsection{{{day}}}\n"
    psalterium += f"\\label{{sec:{short_day}-perannum}}"
    psalterium += f"\\gregorioscore{{../psalm-antiphons/{day.lower().replace(' ','')}antiphon.gabc}}\n\n"
    

    for i, psalm in enumerate(psalms):
        i += 1
        try:
            a, b = str(psalm).split(".")
            psalterium += f"\\ps[{psalm_parts[b]}]{{{a}}}\n"
        except ValueError:
            psalterium += f"\\ps{{{psalm}}}\n"

        psalterium += f"\\gregorioscore{{perannum/{short_day}{i}.gabc}}\n"
        psalterium += "\n\\begin{verse}\n"
        with Path(f"perannum/{short_day}{i}.tex").open() as f:
            psalterium += f.read()
        psalterium += "\end{verse}\n"
    
        # psalterium += f"\\input{{perannum/{short_day}{i}.tex}}\n"

    psalterium += "\n\n"

psalterium += '\subsection{Tempore Paschalis}'
psalterium += '\let\subsection\\tmp'
psalterium += '\gregorioscore{../psalm-antiphons/alleluiapaschal.gabc}'

for day, psalms in days.items():
    short_day = day.split(" ")[-1].lower()[:3]
    psalterium += f"\\subsection{{{day}}}\n"
    psalterium += f"\\label{{sec:{short_day}-paschal}}"
    

    for i, psalm in enumerate(psalms):
        i += 1
        use_previous = None
        with Path(f"perannum/{short_day}1.gabc").open() as f:
            for line in f:
                if line.startswith("annotation"):
                    if '8' in line:
                        use_previous = True
                    else:
                        use_previous = False
                    break

        if use_previous is None:
            print("Unable to work out mode")

        if use_previous:
            psalterium += f"See p. \\pageref{{sec:{short_day}-perannum}}."
            break
                
        try:
            a, b = str(psalm).split(".")
            psalterium += f"\\ps[{psalm_parts[b]}]{{{a}}}\n"
        except ValueError:
            psalterium += f"\\ps{{{psalm}}}\n"

        psalterium += f"\\gregorioscore{{paschal/{short_day}{i}.gabc}}\n"
        psalterium += "\n\\begin{verse}\n"
        with Path(f"paschal/{short_day}{i}.tex").open() as f:
            psalterium += f.read()
        psalterium += "\end{verse}\n"

    psalterium += "\n\n"



with output.open("w") as f:
    f.write(psalterium)


print(psalterium)
        

        
        
        


compline/psalms.py
#!/usr/bin/python3
from os import path
import re
"""Read and parse files from divinumofficium.org to spew out the right psalms."""
"""
[Completorium]
Dominica = Miserére * mihi, Dómine, et exáudi oratiónem meam.
4,90,133
Feria II = Salvum me fac, * Dómine, propter misericórdiam tuam.
6,7(2-10),7(11-18)
Feria III = Tu, Dómine, * servábis nos: et custódies nos in ætérnum.
11,12,15
Feria IV = Immíttet Ángelus Dómini * in circúitu timéntium eum: et erípiet eos.
33(2-11),33(12-23),60
Feria V = Adjútor meus * et liberátor meus esto, Dómine.
69,70(1-12),70(13-24)
Feria VI = Voce mea * ad Dóminum clamávi: neque obliviscétur miseréri Deus.
76(2-13),76(14-21),85
Sabbato = Intret orátio mea * in conspéctu tuo, Dómine.
87,102(1-12),102(13-22)
"""
divinumofficium = path.expanduser(
    "~/code/divinum-officium/web/www/horas/Latin")

rubric = {}
entry = []
# get data
with open(divinumofficium + "/Psalterium/Psalmi minor.txt", "r") as f:
    for line in f:
        if line.startswith("["):
            if len(entry) > 0:  # old entry
                rubric[entry_title] = entry
            entry = []
            entry_title = line.strip("[]\n")
        else:
            if len(line.strip()) > 0:
                entry.append(line.strip())

c = rubric['Completorium']  # this destroys in-situ
while c:
    (day, rep), psalms = (c[0].split("=")), c[1].split(",")
    c[:2] = []
    print("\\subsection{"+ day.strip() + "}\n")
    print(rep.strip(), "\n")
    for psalm in psalms:
        if "(" in psalm:
            (psalm, chunk) = psalm.split("(")
            chunk = chunk.strip(")").split("-")
        else:
            chunk = False
        print("\\begin{verse}")

        with open(divinumofficium + "/psalms1/Psalm" + psalm.strip() + ".txt",
                  "r") as f:
            lines = []
            start = False
            stop = False
            for line in f:
                if stop:
                    break
                if chunk:
                    if not start and psalm.strip() + ":" + chunk[0] in line:
                        start = True
                    elif start and start and psalm.strip() + ":" + chunk[1] in line:
                        stop = True
                else:
                    start = True
                if not start:
                    continue
                line = re.sub(r'[0-9]+:[0-9]+ ', '', line.strip()).replace(
                    "*", "\\*\\\\\n\\vin")
                lines.append(line + "\\\\")
            lines[-1] = lines[-1].rstrip("\\\\")
            for line in lines:
                print(line)
        print("\\end{verse}\n")
    print("\n" + rep.strip() + "\n")
